test/integration/emulator/TimeTurner.cpp

```diff
@@ -0,0 +1,120 @@
+#include "TimeTurner.h"
+
+namespace emulator
+{
+
+TimeTurner::TimeTurner()
+    : _timestamp(100),
+      _startTime(std::chrono::system_clock::now()),
+      _running(true),
+      _sleepersCount(0)
+{
+}
+
+void TimeTurner::nanoSleep(const uint64_t nanoSeconds)
+{
+    if (!_running)
+    {
+        utils::Time::rawNanoSleep(nanoSeconds);
+        return;
+    }
+
+    _sleepersCount.fetch_add(1);
+
+    const auto timerExpires = _timestamp.load() + nanoSeconds;
+    for (size_t i = 0; i < _sleepers.size() * 5; ++i)
+    {
+        auto& slot = _sleepers[i % _sleepers.size()];
+        auto expectedState = State::Empty;
+        if (slot.state.compare_exchange_strong(expectedState, State::Allocated))
+        {
+            slot.expireTimestamp = timerExpires;
+            slot.state.store(State::Committed);
+
+            slot.semaphore.wait();
+            slot.state.store(State::Empty);
+            _sleepersCount.fetch_sub(1);
+            return;
+        }
+    }
+    assert(false);
+}
+
+std::chrono::system_clock::time_point TimeTurner::wallClock()
+{
+    return _startTime +
+        std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::nanoseconds(_timestamp.load()));
+}
+
+/**
+ * Requires all threads to be created and asleep. You cannot remove or add threads.
+ */
+void TimeTurner::runFor(uint64_t nanoSeconds)
+{
+    const auto startTime = getAbsoluteTime();
+
+    const auto threadCount = _sleepersCount.load();
+    while (getAbsoluteTime() - startTime < nanoSeconds)
+    {
+        advance();
+        do
+        {
+            utils::Time::rawNanoSleep(10);
+        } while (_sleepersCount.load() < threadCount);
+    }
+}
+
+void TimeTurner::advance()
```

Comment:
I will change this to wait for a sempahore until threads have completed. It allows for faster progress.

Reply:
+1